File: android/app/src/main/java/agilelens/understudy/teleprompter/ScenePlacer.kt
package agilelens.understudy.teleprompter

import agilelens.understudy.model.Cue
import agilelens.understudy.model.Id
import agilelens.understudy.model.Mark
import agilelens.understudy.model.Pose
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.max
import kotlin.math.sin
import kotlin.math.sqrt

/**
 * Kotlin port of Swift's ScenePlacer. Given a scene from a bundled play
 * and an origin pose, generate a zig-zag walk of Marks with the scene's
 * dialogue bucketed by speaker (max 4 lines per beat). Stage directions
 * attach as leading/trailing .note cues on the beat's Mark.
 */
object ScenePlacer {

    private const val MAX_LINES_PER_BEAT = 4

    fun layout(
        scene: PlayScript.Scene,
        origin: Pose,
        spacing: Float = 1.2f,
        lateralOffset: Float = 0.8f,
        sequenceOffset: Int = 0,
    ): List<Mark> {
        val beats = bucket(scene.entries)

        // Author's forward direction (phone's yaw) = (sin, 0, -cos).
        val forwardX = sin(origin.yaw)
        val forwardZ = -cos(origin.yaw)
        // Right vector = (cos, 0, sin).
        val rightX = cos(origin.yaw)
        val rightZ = sin(origin.yaw)

        val out = mutableListOf<Mark>()
        for ((i, beat) in beats.withIndex()) {
            val side = if (i % 2 == 0) -1f else 1f
            val step = (i + 1) * spacing
            val posX = origin.x + forwardX * step + rightX * side * lateralOffset
            val posZ = origin.z + forwardZ * step + rightZ * side * lateralOffset

            // Yaw toward the next beat (or center for the last beat).
            val yaw: Float = if (i < beats.size - 1) {
                val nextSide = if ((i + 1) % 2 == 0) -1f else 1f
                val nextX = origin.x + forwardX * (i + 2) * spacing +
                    rightX * nextSide * lateralOffset
                val nextZ = origin.z + forwardZ * (i + 2) * spacing +
                    rightZ * nextSide * lateralOffset
                val dx = nextX - posX
                val dz = nextZ - posZ
                atan2(dx, -dz)
            } else {
                // Point back toward the spine — away from the lateral offset.
                atan2(-rightX * side * lateralOffset, rightZ * side * lateralOffset)
            }

            out.add(Mark(
                id = Id(),
                name = beat.title,
                pose = Pose(x = posX, y = 0f, z = posZ, yaw = yaw),
                radius = 0.6f,
                cues = beat.cues(),
                sequenceIndex = sequenceOffset + i,
            ))
        }
        return out
    }

    private data class Beat(
        val leadingStage: MutableList<String> = mutableListOf(),
        var speaker: String = "",
        val lines: MutableList<Pair<String, String>> = mutableListOf(),  // (character, text)
        val trailingStage: MutableList<String> = mutableListOf(),
    ) {
        val title: String
            get() = when {
                lines.isNotEmpty() -> lines[0].first.replaceFirstChar { it.uppercase() }
                leadingStage.isNotEmpty() ->
                    leadingStage.first().let { if (it.length > 24) it.take(23) + "…" else it }
                else -> "Beat"
            }

        fun cues(): List<Cue> = buildList {
            for (stage in leadingStage) {
                add(Cue.Note(id = Id(), text = stage))
            }
            for ((character, text) in lines) {
                add(Cue.Line(id = Id(), text = text, character = character))
            }
            for (stage in trailingStage) {
                add(Cue.Note(id = Id(), text = stage))
            }
        }
    }

    private fun bucket(entries: List<PlayScript.Entry>): List<Beat> {
        val raw = mutableListOf<Beat>()
        var current = Beat()

        fun flush() {
            if (current.lines.isNotEmpty() || current.leadingStage.isNotEmpty()) {
                raw.add(current)
            }
            current = Beat()
        }

        for (entry in entries) {
            when (entry) {
                is PlayScript.Entry.Stage -> {
                    if (current.lines.isEmpty()) current.leadingStage.add(entry.text)
                    else current.trailingStage.add(entry.text)
                }
                is PlayScript.Entry.Line -> {
                    if (current.lines.isEmpty()) {
                        current.speaker = entry.character
                        current.lines.add(entry.character to entry.text)
                    } else if (entry.character == current.speaker &&
                        current.lines.size < MAX_LINES_PER_BEAT
                    ) {
                        current.lines.add(entry.character to entry.text)
                    } else {
                        flush()
                        current.speaker = entry.character
                        current.lines.add(entry.character to entry.text)
                    }
                }
            }
        }
        flush()

        // Merge leading-stage-only beats into the next beat so every output
        // mark has at least one line (unless the whole scene has no lines).
        val merged = mutableListOf<Beat>()
        val carry = mutableListOf<String>()
        for (beat in raw) {
            if (beat.lines.isEmpty()) {
                carry.addAll(beat.leadingStage)
                carry.addAll(beat.trailingStage)
            } else {
                val b = beat.copy(
                    leadingStage = (carry + beat.leadingStage).toMutableList()
                )
                carry.clear()
                merged.add(b)
            }
        }
        if (carry.isNotEmpty() && merged.isNotEmpty()) {
            merged.last().trailingStage.addAll(carry)
        } else if (carry.isNotEmpty()) {
            merged.add(Beat(leadingStage = carry.toMutableList()))
        }
        return merged
    }
}

File: Understudy/Shared/ScenePlacer.swift
//
//  ScenePlacer.swift
//  Understudy
//
//  Given a Scene from the bundled script and an origin pose in the room,
//  generate a natural arrangement of Marks with the scene's dialogue
//  distributed across them. One "Drop Whole Scene" tap produces a
//  walkable staged blocking.
//
//  Placement algorithm:
//    - Walk the scene's entries in order, grouping consecutive entries
//      (lines + inline stage directions) into "beats" when a speaker
//      changes, an exit/entrance happens, or the beat exceeds ~4 lines.
//    - Arrange beats along a gentle zig-zag path in front of the author,
//      spaced ~1.2 m apart, offset alternating left/right by ~0.8 m.
//    - Each beat becomes one Mark: name = "<CHARACTER> (beat N)",
//      cues = any stage direction as a .note + the lines as .line cues.
//    - Preamble stage directions (before the first line) go on mark 1
//      as leading notes.
//

import Foundation
import simd

public enum ScenePlacer {

    /// Public entry: given a scene and a starting pose, return a set of
    /// freshly-built Marks ready to insert into a BlockingStore.
    public static func layout(
        scene: PlayScript.Scene,
        origin: Pose,
        spacing: Float = 1.2,
        lateralOffset: Float = 0.8,
        sequenceOffset: Int = 0
    ) -> [Mark] {
        // Direction the author is facing when they tap "Drop Whole Scene."
        // Marks walk forward from that direction, zig-zagging side to side.
        let forward = SIMD3<Float>(sin(origin.yaw), 0, -cos(origin.yaw))
        let right = SIMD3<Float>(cos(origin.yaw), 0, sin(origin.yaw))

        let beats = bucket(entries: scene.entries)

        var marks: [Mark] = []
        for (i, beat) in beats.enumerated() {
            let side: Float = (i % 2 == 0) ? -1 : 1
            let step = forward * (Float(i + 1) * spacing)
            let lateral = right * side * lateralOffset
            let origin3 = SIMD3<Float>(origin.x, 0, origin.z)
            let position = origin3 + step + lateral

            // Face the next beat (or scene center if last).
            let yaw: Float
            if i < beats.count - 1 {
                let nextStep = forward * Float(i + 2) * spacing
                let nextLat = right * -side * lateralOffset
                let nextPos = origin3 + nextStep + nextLat
                let d = nextPos - position
                yaw = atan2f(d.x, -d.z)
            } else {
                yaw = atan2f(-lateral.x, lateral.z)
            }

            let mark = Mark(
                id: ID(),
                name: beat.title,
                pose: Pose(x: position.x, y: 0, z: position.z, yaw: yaw),
                radius: 0.6,
                cues: beat.cues(),
                sequenceIndex: sequenceOffset + i
            )
            marks.append(mark)
        }

        return marks
    }

    // MARK: - Beat bucketing

    /// One "beat" — a speaker's turn, possibly spanning multiple consecutive
    /// line entries plus inline stage directions. Capped at ~4 dialogue
    /// entries to keep beats walkable.
    private struct Beat {
        var leadingStage: [String] = []      // stage directions before the first line
        var speaker: String = ""
        var lines: [(text: String, character: String)] = []
        var trailingStage: [String] = []

        var title: String {
            if lines.isEmpty {
                return leadingStage.first.map { shortened($0, to: 24) } ?? "Beat"
            }
            let speakerLabel = lines[0].character
            return "\(speakerLabel.capitalized)"
        }

        func cues() -> [Cue] {
            var result: [Cue] = []
            for stage in leadingStage {
                result.append(.note(id: ID(), text: stage))
            }
            for l in lines {
                result.append(.line(id: ID(), text: l.text, character: l.character))
            }
            for stage in trailingStage {
                result.append(.note(id: ID(), text: stage))
            }
            return result
        }
    }

    private static let maxLinesPerBeat = 4

    private static func bucket(entries: [PlayScript.Entry]) -> [Beat] {
        // Step 1: build raw beats, one per speaker turn (or preamble-stage).
        var beats: [Beat] = []
        var current = Beat()

        func flush() {
            if !current.lines.isEmpty || !current.leadingStage.isEmpty {
                beats.append(current)
            }
            current = Beat()
        }

        for entry in entries {
            switch entry {
            case .stage(let text):
                if current.lines.isEmpty {
                    current.leadingStage.append(text)
                } else {
                    current.trailingStage.append(text)
                }
            case .line(let character, let text, _):
                if current.lines.isEmpty {
                    current.speaker = character
                    current.lines.append((text: text, character: character))
                } else if character == current.speaker,
                          current.lines.count < maxLinesPerBeat {
                    current.lines.append((text: text, character: character))
                } else {
                    flush()
                    current.speaker = character
                    current.lines.append((text: text, character: character))
                }
            }
        }
        flush()

        // Step 2: merge leading-stage-only beats into the next beat, so every
        // output mark has at least one line (unless the whole scene has none).
        var merged: [Beat] = []
        var carryStage: [String] = []
        for beat in beats {
            if beat.lines.isEmpty {
                carryStage.append(contentsOf: beat.leadingStage)
                carryStage.append(contentsOf: beat.trailingStage)
            } else {
                var b = beat
                b.leadingStage.insert(contentsOf: carryStage, at: 0)
                carryStage.removeAll()
                merged.append(b)
            }
        }
        if !carryStage.isEmpty {
            if var last = merged.popLast() {
                last.trailingStage.append(contentsOf: carryStage)
                merged.append(last)
            } else {
                // No dialogue at all — keep a single stage-only beat.
                merged.append(Beat(leadingStage: carryStage))
            }
        }
        return merged
    }

    private static func shortened(_ s: String, to max: Int) -> String {
        s.count <= max ? s : String(s.prefix(max - 1)) + "…"
    }
}
